### PR TITLE
[12.0][FIX] sale: Populate default values + company in virtual invoice for onchanges

### PR DESCRIPTION
**Pull request by pedrobaeza**
If not, you can have side effects on _onchange_partner_id (and overrides), as the company is not set and other possible default values.

Example of problem detected with this code:

https://github.com/OCA/bank-payment/blob/939cb0a2750538ddc28e24446dd9ba5931e95e61/account_payment_partner/models/account_invoice.py#L48

For the default part, I discovered in the past that `new` is not populating default values [1].

@Tecnativa TT25224

[1] https://twitter.com/PedroMBaeza/status/1067086018879397890?s=20

### CODE REVIEW

**Comment by pedrobaeza** [comment]
@dbo-odoo can you take a look to this?

**Comment by pedrobaeza** [comment]
The problem is not per se final values, but possible access problems in a multi-company environment.

**Review by MiquelRForgeFlow — APPROVED** [review]

File: addons/sale/models/sale.py
Comment (Feyensv) [inline addons/sale/models/sale.py:482]:
```suggestion
        vinvoice = self.env['account.invoice'].new({
            'partner_id': self.partner_invoice_id.id,
            'type': 'out_invoice',
            'company_id': company_id})
```
Updating all the default values may have unknown side-effects and be risky for a stable branch, but I can accept the `company_id` value as a targeted fix for multi-company purposes.

**Comment by pedrobaeza** [comment]
Thanks for the review, @Feyensv.

Although it seems a bit hard to have full default dictionary, being only assigned at the beginning and overwritten by the onchanges and rest of the logic, it's not very likely to provoke any error, and on contrary can have other side effects if other code relies on a default value not populated. Are you sure this can't be put as is?

Don't get me wrong, I prefer to have it merged at least for the company one, but if we advance any possible potential problem with the more complete patch, it's better IMO.

**Comment by Feyensv** [comment]
I took some time to investigate the problem, and IMHO, this virtual invoice creation shouldn't happen here at all.
There is a similar onchange simulation in the `account.invoice` `create` method, and we should leave the management of onchanges to it, instead of doing it twice (and the first time with incomplete invoice information).

And for your custom OCA module, in 12.0, you should hook yourself on `account.invoice`.`_get_onchange_create()` method, to extend the fields to specify with `_onchange_partner_id`, instead of doing the same virtual invoice creation ;) (Cf https://github.com/odoo/odoo/commit/5eb40654865d3e04526d98de2a10c0c02dde0c0c).

**Comment by pedrobaeza** [comment]
Thanks for the investigation. Then should I change this PR to remove all the onchange stuff?

**Comment by Feyensv** [comment]
I'll make it in a separate PR for the moment, and see if I can get a review from sbdy else to be sure I'm not loosing sthg :).

**Comment by pedrobaeza** [comment]
Thanks, I'll check it as well to give feedback and adapt the OCA module according it. I keep this PR opened for now while not checked if you don't mind.

**Comment by Feyensv** [comment]
No problem, I'll close this PR if/once my fix is merged ;).

**Comment by Feyensv** [comment]
> Thanks, I'll check it as well to give feedback and adapt the OCA module according it. I keep this PR opened for now while not checked if you don't mind.

#68136 was merged, I believe it correctly fixes your problem but since you didn't confirm it, can you check it works fine now ? :).

**Comment by pedrobaeza** [comment]
Hi, @Feyensv, sorry for the late reply. I haven't dug in deep, but no issues reported from any customer, so I suppose it's OK (although we still feel the pain of not having consistent results due to the duality of onchanges/other logic, and need to provide glue modules when expanding the possible fields in invoices). Closing this PR thus.

Thanks for your attention.

**Pull request closed without merging**
